src/test/java/com/google/neighborgood/UserInfoServletTest.java

```diff
@@ -0,0 +1,422 @@
+    dummy.setProperty("nickname", "Leonard");
+    dummy.setProperty("address", "xxx");
+    dummy.setProperty("phone", "xxx");
+    dummy.setProperty("email", "test@example.com");
+    dummy.setProperty("userId", "1234567890");
+    dummy.setProperty("points", 0);
+    ds.put(dummy);
+
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    try {
+      new UserInfoServlet().doGet(request, response);
+    } catch (IOException e) {
+      assertTrue(false);
+    }
+
+    // After sending the GET request, the doGet function should output the json string
+    writer.flush();
+    assertTrue(stringWriter.toString().contains("[\"Leonard\",\"xxx\",\"xxx\"]"));
+  }
+
+  @Test
+  public void stressTest() {
+    HttpServletRequest request = mock(HttpServletRequest.class);
+    HttpServletResponse response = mock(HttpServletResponse.class);
+
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+
+    // Insert 999 dummy entities into datastore to simulate 999 users
+    for (int i = 0; i < 999; i++) {
+      String str = Integer.toString(i);
+      Entity dummy = new Entity("UserInfo");
+      dummy.setProperty("nickname", "Leonard" + str);
+      dummy.setProperty("address", "xxx" + str);
+      dummy.setProperty("phone", str + str);
+      dummy.setProperty("email", "test" + str + "@example.com");
+      dummy.setProperty("userId", str);
+      dummy.setProperty("points", 0);
+      ds.put(dummy);
+    }
+
+    // Check the number of entities in the datastore. The result should be 999
+    assertEquals(999, ds.prepare(new Query("UserInfo")).countEntities());
+
+    // Now we insert a target UserInfo Entity and check the time needed to find the entity
+    when(request.getParameter("nickname-input")).thenReturn("Leonard");
+    when(request.getParameter("address-input")).thenReturn("4xxx Centre Avenue");
+    when(request.getParameter("phone-input")).thenReturn("4xxxxxxxxx");
+
+    try {
+      new UserInfoServlet().doPost(request, response);
+    } catch (IOException e) {
+      assertTrue(false);
+    }
+
+    // After sending the POST request, there should be three entities in the datastore
+    assertEquals(1000, ds.prepare(new Query("UserInfo")).countEntities());
+
+    Instant start = Instant.now();
+
+    // Filter out the entity that has the userId of 1234567890
+    PreparedQuery results =
+        ds.prepare(
+            new Query("UserInfo")
+                .setFilter(new FilterPredicate("userId", FilterOperator.EQUAL, "1234567890")));
+    Entity entity = results.asSingleEntity();
+
+    Instant end = Instant.now();
+    Duration timeElapsed = Duration.between(start, end);
+
+    // Fetching the specific user info from datastore should take a reasonable amount of time
+    // I think 10ms is a reasonable threshold value.
+    assertTrue(timeElapsed.toMillis() < 10);
```

Comment:
While it's good to track the performance of our site, we should really attempt to make our tests stable. Furthermore, this is testing the performance of a local version of a dependency, which isn't really under our control. We can trust the Datastore helper to be reasonably performant in our unit tests.

This also brings up a question for me - have you set up indexes on the columns (like userId) that you intend to query frequently? If you do that, you can expect Datastore to perform adequately in these lookups, regardless of the amount of data.

If you do want to add an end-to-end test with junit, we could tag it and create a different Maven target that includes tests with the tag you choose. Then it won't run this test every time, which will be nice while developing the site, since those kinds of tests can be flaky and/or time consuming. See https://www.baeldung.com/junit-filtering-tests, for example.

Reply:
Ah I actually have not set up indexes on the columns on the userId. But it's really a great idea. After setting it up, most queries should be handled really efficiently. I will record this as an issue and implemented it after the MVP.

And as discussed in chat yesterday, I decided to remove the stress tests for now. But I will look at the JUnit filtering test later on to implement a stable stress test suite for the servlet files.